go/ql/test/library-tests/semmle/go/frameworks/Fasthttp/fasthttp.go

```diff
@@ -0,0 +1,225 @@
+package main
+
+// go:generate depstubber -vendor  "github.com/valyala/fasthttp" Args,Client,Cookie,FS,HostClient,LBClient,PathRewriteFunc,Request,RequestCtx,RequestHandler,RequestHeader,Response,ResponseHeader,Server,TCPDialer,URI,LBClient,PipelineClient AcquireURI,Serve,DialDualStack,Dial,DialTimeout,DialDualStackTimeout,Get,GetDeadline,GetTimeout,Post,Do,DoRedirects,AppendHTMLEscapeBytes,AppendHTMLEscape,AppendQuotedArg,ServeFileBytesUncompressed,ServeFileBytes,ServeFileUncompressed,ServeFile,SaveMultipartFile,DoTimeout,DoDeadline
+
+import (
+	"bufio"
+	"net"
+	"time"
+
+	"github.com/valyala/fasthttp"
+)
+
+func source() interface{} {
+	return make([]byte, 1)
+}
+
+func sink(interface{}) {
+}
+
+func fasthttpClient() {
+	userInput := "127.0.0.1:8909"
+	fasthttp.DialDualStack(userInput)           // $ SsrfSink=userInput
+	fasthttp.Dial(userInput)                    // $ SsrfSink=userInput
+	fasthttp.DialTimeout(userInput, 5)          // $ SsrfSink=userInput
+	fasthttp.DialDualStackTimeout(userInput, 5) // $ SsrfSink=userInput
+
+	res := &fasthttp.Response{}
+	req := &fasthttp.Request{}
+	req.SetHost(source().(string))
+	sink(req) // $ hasTaintFlow="req"
+	req.SetHostBytes(source().([]byte))
+	sink(req) // $ hasTaintFlow="req"
+	req.SetRequestURI(source().(string))
+	sink(req) // $ hasTaintFlow="req"
+	req.SetRequestURIBytes(source().([]byte))
+	sink(req) // $ hasTaintFlow="req"
```

Comment:
You still have the problem that once `req` is tainted, it never loses the taint. You really need to use different variables.
```suggestion
	req1 := &fasthttp.Request{}
	req1.SetHost(source().(string))
	sink(req1) // $ hasTaintFlow="req1"
	req2 := &fasthttp.Request{}
	req2.SetHostBytes(source().([]byte))
	sink(req2) // $ hasTaintFlow="req2"
	req3 := &fasthttp.Request{}
	req3.SetRequestURI(source().(string))
	sink(req3) // $ hasTaintFlow="req3"
	req4 := &fasthttp.Request{}
	req4.SetRequestURIBytes(source().([]byte))
	sink(req4) // $ hasTaintFlow="req4"
```
Please do the same for `uri`.